From patchwork Sat Oct 25 04:31:24 2025
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: iuncuim <iuncuim@gmail.com>
X-Patchwork-Id: 14284213
Return-Path: 
 <linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from bombadil.infradead.org (bombadil.infradead.org
 [198.137.202.133])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by smtp.lore.kernel.org (Postfix) with ESMTPS id 97AB4CCF9E3
	for <linux-arm-kernel@archiver.kernel.org>;
 Sat, 25 Oct 2025 04:32:21 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=bombadil.20210309; h=Sender:List-Subscribe:List-Help
	:List-Post:List-Archive:List-Unsubscribe:List-Id:Content-Transfer-Encoding:
	MIME-Version:References:In-Reply-To:Message-ID:Date:Subject:Cc:To:From:
	Reply-To:Content-Type:Content-ID:Content-Description:Resent-Date:Resent-From:
	Resent-Sender:Resent-To:Resent-Cc:Resent-Message-ID:List-Owner;
	bh=CuZHP6YFsCwwA4TkyjhYi+XmEZKmSYURob9jZc/v+s0=; b=G+z2E55BMkzV/nqbG/ww9kWR3E
	1ApeUShBQkBEtdk3CMeeytr1YwN4Bd5pGhotpMqx3lTrqhx1Ni6UbakEIcRtsxNfbB+J1IqnAW1rM
	rhOIuzWDQeZSnUy4I/rQVg+WTqDm76dpmTR6bZoSVBN3Fe79YlideOIsxZkEuK4+jslTsh20dzVkw
	0CUVxLoUWaIIcMfLXde0ZsUyyDL7K8OQhRS0daSArRtcwIlDTeWb06Kifkp6KQ2EpBmLJxpXFq76c
	j1fkUS7k4AwLZe2vFOuZ58dqxiYsr0RBVI4ZcmXAE0On/un+yA0niDUzyYsQdcH7nWPDF0mQeGMLm
	T/gnkqjA==;
Received: from localhost ([::1] helo=bombadil.infradead.org)
	by bombadil.infradead.org with esmtp (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVwt-0000000AwSQ-1zqb;
	Sat, 25 Oct 2025 04:32:15 +0000
Received: from mail-lj1-x22c.google.com ([2a00:1450:4864:20::22c])
	by bombadil.infradead.org with esmtps (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVwr-0000000AwQz-0ZUB
	for linux-arm-kernel@lists.infradead.org;
	Sat, 25 Oct 2025 04:32:14 +0000
Received: by mail-lj1-x22c.google.com with SMTP id
 38308e7fff4ca-3737d0920e6so44916821fa.1
        for <linux-arm-kernel@lists.infradead.org>;
 Fri, 24 Oct 2025 21:32:12 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20230601; t=1761366731; x=1761971531;
 darn=lists.infradead.org;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:from:to:cc:subject:date
         :message-id:reply-to;
        bh=CuZHP6YFsCwwA4TkyjhYi+XmEZKmSYURob9jZc/v+s0=;
        b=gJBdJLAiTPWeG/1DD7EuhlYSnw3qMuGzWNkEuiZThNagiVpgqzKsacMa9ebr0hAI43
         G/ngshtUT9OcGkmC0qt/tdWDq8XidWx8zb8sBxYxrNzyMT/4NYU6kWeBFLoh60aiKrwK
         EqfldXb6crbnhNHBxUBIkKMEB8ANtt7hUkuGrs1tsEektX7zCErW4m7ko5cBCLQe4KkQ
         NbcNkTQXwKLLxaE6k2SmZxnFg9RvmN2VN2s3sHplK/G+iZXX2il1c6RH453Fb13xXVbS
         /T59Yp443oktjIEcoyHB6hYPNY0vN2A3Hmhp0MVLccT2FoMMz0k2l6PnLTe+Eivid5Lo
         LXNw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20230601; t=1761366731; x=1761971531;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:x-gm-message-state:from:to:cc
         :subject:date:message-id:reply-to;
        bh=CuZHP6YFsCwwA4TkyjhYi+XmEZKmSYURob9jZc/v+s0=;
        b=OzaLGwTTZaRrqVtEIZlLn4ZGl5hY/YJJgqE/cvIb/7AOwpTnWaRXzBZ6dEnpRpkTCg
         1OgAqquq/zyd3f958VsMyQCf/VzYQgdJAQTIKFCONrrYa7DwFOr2oDRh/WiN3M521aLR
         XymemCKOhgbhyIT3bgBjuc9OUYmArJlEVzFp3A7MTPfaRyadgreIcQTQpLeX27DB/FNM
         LAcT4vYCgLvnc93LNvWMZYzsYZ2lmglUee0/chE98mYJHawlZbinwONNqxRqgf5bxAli
         cddJ0qzANWkk/HaewfH1Hn/nNfZctGFSEo7JmBiEq84a2FYbIr+QJvaHCboGYGV80cfc
         voog==
X-Forwarded-Encrypted: i=1;
 AJvYcCUHXUNshHc2RyYwpiXq8/gdhC891OUBVtL1M+IWV2he4p61ZBoPhPkc2eAJc+Y7J10Zh2yOGNogdK6pFb7RB2Xf@lists.infradead.org
X-Gm-Message-State: AOJu0Yz7KeQldJP4OSESsrnBWttvG8m516gvBZbAUBKbAjISIs1Ba67E
	UGjQ6CQItjUERLB2TBI8VIkyPPyYHk2EMhZZCVqTLdzS4raLsUsYQkXm
X-Gm-Gg: ASbGncuCilPo48EdP67TC7749/wcDqEWWBEvVzv1hhVJko9ZHYzKFMmfPDKVHb8QFra
	pVX0aTSFO5JhD3GO4SlB2qnorCV8TAOZX7vOy75VlJxIpSSPhbeCJSpLCoiNT1wHdcnlnzAxy21
	MUxXzHn8dIAWErIQ9Yj8TTxJl7eAQ0deL0pjXQVrn4HmytdZIxTZWruRvM664+vAM3C5UuaGfpC
	FiWr1lHaea36LVExSGrRgmUaDpqCSctgGMjkZ1wk+y5jqWFx9Y8Fn//QrJB3wxMendg5UChRLkL
	g4AFohRsW0AAi1xiGjyJiid87U8FXb92jhln1Uh+Mc/XKUQuUqr66kWVyJaZBU482Cb8xEIKVCw
	xlmfWXB+CRNz8K6O7teCs+Yl1sNJW6YYElWQFxOTygAbCQBp0r3Zr3KF9sVo/6dtEBlSOjImxJx
	C9
X-Google-Smtp-Source: 
 AGHT+IHidqK20asSnkSoaYA0Ua+IEXaUMgtNo1d+OXf+4JbTCbl0aR8F8Cd5djjk9glXI6+3c84fFA==
X-Received: by 2002:a2e:a916:0:b0:376:4430:b545 with SMTP id
 38308e7fff4ca-378e4648542mr13761041fa.49.1761366730866;
        Fri, 24 Oct 2025 21:32:10 -0700 (PDT)
Received: from junAIR ([176.106.241.81])
        by smtp.gmail.com with ESMTPSA id
 38308e7fff4ca-378ee0ca7a0sm3409241fa.33.2025.10.24.21.31.55
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Fri, 24 Oct 2025 21:32:10 -0700 (PDT)
From: iuncuim <iuncuim@gmail.com>
To: Vasily Khoruzhick <anarsoul@gmail.com>,
	Yangtao Li <tiny.windzz@gmail.com>,
	"Rafael J. Wysocki" <rafael@kernel.org>,
	Daniel Lezcano <daniel.lezcano@linaro.org>,
	Zhang Rui <rui.zhang@intel.com>,
	Lukasz Luba <lukasz.luba@arm.com>,
	Rob Herring <robh@kernel.org>,
	Krzysztof Kozlowski <krzk+dt@kernel.org>,
	Conor Dooley <conor+dt@kernel.org>,
	Chen-Yu Tsai <wens@csie.org>,
	Jernej Skrabec <jernej.skrabec@gmail.com>,
	Samuel Holland <samuel@sholland.org>,
	Philipp Zabel <p.zabel@pengutronix.de>
Cc: Andre Przywara <andre.przywara@arm.com>,
	linux-pm@vger.kernel.org,
	devicetree@vger.kernel.org,
	linux-arm-kernel@lists.infradead.org,
	linux-sunxi@lists.linux.dev,
	linux-kernel@vger.kernel.org
Subject: [PATCH v3 1/6] dt-bindings: thermal: sun8i: Add A523 THS0/1
 controllers
Date: Sat, 25 Oct 2025 12:31:24 +0800
Message-ID: <20251025043129.160454-2-iuncuim@gmail.com>
X-Mailer: git-send-email 2.51.0
In-Reply-To: <20251025043129.160454-1-iuncuim@gmail.com>
References: <20251025043129.160454-1-iuncuim@gmail.com>
MIME-Version: 1.0
X-CRM114-Version: 20100106-BlameMichelson ( TRE 0.8.0 (BSD) ) MR-646709E3 
X-CRM114-CacheID: sfid-20251024_213213_205585_557D90E4 
X-CRM114-Status: GOOD (  13.15  )
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.34
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: 
 <http://lists.infradead.org/mailman/options/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: 
 <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
Sender: "linux-arm-kernel" <linux-arm-kernel-bounces@lists.infradead.org>
Errors-To: 
 linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org

From: Mikhail Kalashnikov <iuncuim@gmail.com>

Add a binding for D1/T113s thermal sensor controller. Add dt-bindings
description of the thermal sensors in the A523 processor.
The controllers require activation of the additional frequency of the
associated gpadc controller, so a new clock property has been added.

The calibration data is split into two cells that are in different areas
of nvmem. Both controllers require access to both memory cell, so a new
property nvmem-cells has been added. To maintain backward compatibility,
the name of the old cell remains the same and the new nvmem-cell-names is
called calibration-second-part

Signed-off-by: Mikhail Kalashnikov <iuncuim@gmail.com>
---
 .../thermal/allwinner,sun8i-a83t-ths.yaml     | 56 ++++++++++++++++++-
 1 file changed, 53 insertions(+), 3 deletions(-)

diff --git a/Documentation/devicetree/bindings/thermal/allwinner,sun8i-a83t-ths.yaml b/Documentation/devicetree/bindings/thermal/allwinner,sun8i-a83t-ths.yaml
index 3e61689f6..b2f750ef2 100644
--- a/Documentation/devicetree/bindings/thermal/allwinner,sun8i-a83t-ths.yaml
+++ b/Documentation/devicetree/bindings/thermal/allwinner,sun8i-a83t-ths.yaml
@@ -24,18 +24,23 @@ properties:
       - allwinner,sun50i-h5-ths
       - allwinner,sun50i-h6-ths
       - allwinner,sun50i-h616-ths
+      - allwinner,sun55i-a523-ths0
+      - allwinner,sun55i-a523-ths1
 
   clocks:
     minItems: 1
     items:
       - description: Bus Clock
       - description: Module Clock
+      - description: GPADC Clock
 
   clock-names:
     minItems: 1
+    maxItems: 2
     items:
       - const: bus
       - const: mod
+      - const: gpadc
 
   reg:
     maxItems: 1
@@ -47,11 +52,16 @@ properties:
     maxItems: 1
 
   nvmem-cells:
-    maxItems: 1
-    description: Calibration data for thermal sensors
+    minItems: 1
+    items:
+      - description: Calibration data for thermal sensors
+      - description: Additional cell in case of separate calibration data
 
   nvmem-cell-names:
-    const: calibration
+    minItems: 1
+    items:
+      - const: calibration
+      - const: calibration-second-part
 
   allwinner,sram:
     maxItems: 1
@@ -107,6 +117,7 @@ allOf:
             enum:
               - allwinner,sun8i-h3-ths
               - allwinner,sun20i-d1-ths
+              - allwinner,sun55i-a523-ths0
 
     then:
       properties:
@@ -132,6 +143,32 @@ allOf:
         - clock-names
         - resets
 
+  - if:
+      properties:
+        compatible:
+          contains:
+            enum:
+              - allwinner,sun55i-a523-ths0
+              - allwinner,sun55i-a523-ths1
+    then:
+      properties:
+        clocks:
+          minItems: 2
+        clock-names:
+          enum: [ bus, gpadc ]
+        nvmem-cells:
+          minItems: 2
+        nvmem-cell-names:
+          minItems: 2
+    else:
+      properties:
+        nvmem-cells:
+          maxItems: 1
+        nvmem-cell-names:
+          maxItems: 1
+          items:
+            - const: calibration
+
 required:
   - compatible
   - reg
@@ -176,4 +213,17 @@ examples:
         #thermal-sensor-cells = <1>;
     };
 
+  - |
+    thermal-sensor@2009400 {
+      compatible = "allwinner,sun55i-a523-ths1";
+      reg = <0x02009400 0x400>;
+      interrupts = <GIC_SPI 62 IRQ_TYPE_LEVEL_HIGH>;
+      clocks = <&ccu CLK_BUS_THS>, <&ccu CLK_GPADC1>;
+      clock-names = "bus", "gpadc";
+      resets = <&ccu RST_BUS_THS>;
+      nvmem-cells = <&ths_calibration0>, <&ths_calibration1>;
+      nvmem-cell-names = "calibration",
+             "calibration-second-part";
+      #thermal-sensor-cells = <1>;
+    };
 ...

From patchwork Sat Oct 25 04:31:25 2025
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: iuncuim <iuncuim@gmail.com>
X-Patchwork-Id: 14284214
Return-Path: 
 <linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from bombadil.infradead.org (bombadil.infradead.org
 [198.137.202.133])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by smtp.lore.kernel.org (Postfix) with ESMTPS id 9C956CCD1BF
	for <linux-arm-kernel@archiver.kernel.org>;
 Sat, 25 Oct 2025 04:32:36 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=bombadil.20210309; h=Sender:List-Subscribe:List-Help
	:List-Post:List-Archive:List-Unsubscribe:List-Id:Content-Transfer-Encoding:
	MIME-Version:References:In-Reply-To:Message-ID:Date:Subject:Cc:To:From:
	Reply-To:Content-Type:Content-ID:Content-Description:Resent-Date:Resent-From:
	Resent-Sender:Resent-To:Resent-Cc:Resent-Message-ID:List-Owner;
	bh=bw4z9PVBnrl74pcQ+V85ZiVa2r5TmKnZAhW/Uv6GfHQ=; b=BFI0gmnYQHKNPr7bJiGmiGk+fP
	D+8/KDVRsMzyBfC13dZ7gNgTys2aPx/dnKK0w9DePZq2T0yCRCG8tOHB+1nFsgNcKnmv+5CK/ZWw2
	8fDv6TLTlPXtHsSZSU+TIcl1PklwRBbT5AvC15yt9QBQl1CXMzggyVDQN1iaow16c55awBEb2mI/h
	zIfsyZ7lgd32ynoY5sVzwV1v0nEKEBM7om+aMCBQO6yzNwlE4yd53o2JN3RyTR1LolbbMGznHT5T7
	OLMmsVK+xrQ/P76HaoLrxjjAPmif8roN2eeE59IeF68OW4wgUSulzfxzaSFybl9zt1NazLxF4lImd
	WNxrA0Dw==;
Received: from localhost ([::1] helo=bombadil.infradead.org)
	by bombadil.infradead.org with esmtp (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVx7-0000000AwWu-3Gku;
	Sat, 25 Oct 2025 04:32:30 +0000
Received: from mail-lj1-x22e.google.com ([2a00:1450:4864:20::22e])
	by bombadil.infradead.org with esmtps (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVx5-0000000AwV4-1ppp
	for linux-arm-kernel@lists.infradead.org;
	Sat, 25 Oct 2025 04:32:28 +0000
Received: by mail-lj1-x22e.google.com with SMTP id
 38308e7fff4ca-378ddffb497so24451231fa.2
        for <linux-arm-kernel@lists.infradead.org>;
 Fri, 24 Oct 2025 21:32:26 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20230601; t=1761366745; x=1761971545;
 darn=lists.infradead.org;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:from:to:cc:subject:date
         :message-id:reply-to;
        bh=bw4z9PVBnrl74pcQ+V85ZiVa2r5TmKnZAhW/Uv6GfHQ=;
        b=UrSEakAPsyKuxmnGcpJo9jMNjWWSsEUMUEHmRXxErDaNOa9mW4JQ5AwKZYyDAOzdCn
         hNvcAq3JT6KF0vI+hT3urIFacPYaieq8/0wLd3oUF5z4h/J9NYAU66S8zJ8qhaWjVQXx
         x8SLkBFSE1zVr/ypyOeMIKkCJU7RELTZeEzITltOn8uAmxfnADC1pB1mt0qeNECGE1yb
         +e/JnXMr8SoSStCmg7oGrCPi3SLqQfynyOz08q7SGDLUO6/MpEnuZcHpZ1mkg5GIJUfa
         fP8rSXabrgRuwVchcM2RsDKK4LJk+3QJVGHNNhEiub/xsrZ9kcHJuOyBw7Bh+8ZQghw/
         6Q5g==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20230601; t=1761366745; x=1761971545;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:x-gm-message-state:from:to:cc
         :subject:date:message-id:reply-to;
        bh=bw4z9PVBnrl74pcQ+V85ZiVa2r5TmKnZAhW/Uv6GfHQ=;
        b=oDDcu9weAoTOhDz+zfuWrqV/xa8ysGcpSng1nXfwZb54PoTu0kLv2ngUKh/JBb6v78
         uDAr1oEuw1VejCXiqtF0wDDbywUwrQwnSdDJURAxLDLpIRq4XfSP/Wy7HoavxaIzm6kg
         LpybRH4cRmXymuYETow+sLm0pFzyQ94i9BiOOiv8YZziT1UEeiYos0IAj8K/d5ilmzu/
         b2uUaBV5nfuy3P2GyJ8pJpMDzdiPLYx1Q/gI7sK9se/hY1rBeMqdiz17+boasf97CP+G
         n7OtSwNTKqfJ8bxJXJ5kkKIHXZYh+2hm32cIwJpVppwIxYveBcZ4Ao03GlDL1UUWfcyt
         tiwA==
X-Forwarded-Encrypted: i=1;
 AJvYcCXppUaao3hTWPIt+pzwQ4t7OP5ZYCqvv494L1hxd4KAa9pr09hFiSUqjax1Ut14z/q/ybg8QIcYNPqk0FTfneq6@lists.infradead.org
X-Gm-Message-State: AOJu0YxNB9DN7z+nLkJ/4qMu/vgQC/dclx9OPg5RwBp6//GMgfXWmr61
	k94RVygYqh/IrKMvn2oEZHV+W3r/A6jEMhn1fyOFGV3OEJZ/y0gxkPq2
X-Gm-Gg: ASbGncui03smJHagKpBH/DqLuXNplAUEjqXEWVi++jmo3w4yKlXR58sU0KSfMigF4yD
	1LZPXpT5sKd60dX+FrZuW9z8mIFx+9c+1x/bF0j5R9IS+BqHuJv3jCcpqnpwh+wSiTsqVcVnlVu
	qh2gCnu/smTytcQj4NyAsGt+du+4lTuwwoWTXpI4UdTR8HJh1zmKAgzx63zKqZ2SsnQma1cZ7AE
	vIwf6a8oUP6pdeLB/bqvLnOG8v12xLvln/1G9iA9bgWEOqkRLwCr4UvCL1zzCRyBrxb6bADfmDE
	vxsZEeAWk7bkAsOpLzBGwFHHYT6NPLzUeJVeQVa8le6DpbwjV7CrtQvDUfk27iwX6LYuTb5jXJ0
	qDfizpH2xJsN+mMLyCjZ5hYUCTo1C5GBzAAzQ3Ejo4k85erz3392pkplygBcgZ9ShA354JrvEce
	QZ
X-Google-Smtp-Source: 
 AGHT+IFIlO0Xnz3NfsASBS+sA6Ogolz52/zgOOHUf1BRiQNX9JIhdrYW1iCIFtP2vXKndrvpB2lUXQ==
X-Received: by 2002:a05:651c:221f:b0:375:db6e:fac9 with SMTP id
 38308e7fff4ca-37797a0dab3mr93162251fa.31.1761366745177;
        Fri, 24 Oct 2025 21:32:25 -0700 (PDT)
Received: from junAIR ([176.106.241.81])
        by smtp.gmail.com with ESMTPSA id
 38308e7fff4ca-378ee0ca7a0sm3409241fa.33.2025.10.24.21.32.11
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Fri, 24 Oct 2025 21:32:24 -0700 (PDT)
From: iuncuim <iuncuim@gmail.com>
To: Vasily Khoruzhick <anarsoul@gmail.com>,
	Yangtao Li <tiny.windzz@gmail.com>,
	"Rafael J. Wysocki" <rafael@kernel.org>,
	Daniel Lezcano <daniel.lezcano@linaro.org>,
	Zhang Rui <rui.zhang@intel.com>,
	Lukasz Luba <lukasz.luba@arm.com>,
	Rob Herring <robh@kernel.org>,
	Krzysztof Kozlowski <krzk+dt@kernel.org>,
	Conor Dooley <conor+dt@kernel.org>,
	Chen-Yu Tsai <wens@csie.org>,
	Jernej Skrabec <jernej.skrabec@gmail.com>,
	Samuel Holland <samuel@sholland.org>,
	Philipp Zabel <p.zabel@pengutronix.de>
Cc: Andre Przywara <andre.przywara@arm.com>,
	linux-pm@vger.kernel.org,
	devicetree@vger.kernel.org,
	linux-arm-kernel@lists.infradead.org,
	linux-sunxi@lists.linux.dev,
	linux-kernel@vger.kernel.org
Subject: [PATCH v3 2/6] thermal/drivers/sun8i: add gpadc clock
Date: Sat, 25 Oct 2025 12:31:25 +0800
Message-ID: <20251025043129.160454-3-iuncuim@gmail.com>
X-Mailer: git-send-email 2.51.0
In-Reply-To: <20251025043129.160454-1-iuncuim@gmail.com>
References: <20251025043129.160454-1-iuncuim@gmail.com>
MIME-Version: 1.0
X-CRM114-Version: 20100106-BlameMichelson ( TRE 0.8.0 (BSD) ) MR-646709E3 
X-CRM114-CacheID: sfid-20251024_213227_493172_47CF3BB8 
X-CRM114-Status: GOOD (  13.03  )
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.34
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: 
 <http://lists.infradead.org/mailman/options/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: 
 <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
Sender: "linux-arm-kernel" <linux-arm-kernel-bounces@lists.infradead.org>
Errors-To: 
 linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org

From: Mikhail Kalashnikov <iuncuim@gmail.com>

Some processors (e.g. Allwinner A523) require GPADC clocking activation for
temperature sensors to work. So let's add support for enabling it.

Signed-off-by: Mikhail Kalashnikov <iuncuim@gmail.com>
Reviewed-by: Chen-Yu Tsai <wens@kernel.org>
---
 drivers/thermal/sun8i_thermal.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/thermal/sun8i_thermal.c b/drivers/thermal/sun8i_thermal.c
index 226747906..c02c398b0 100644
--- a/drivers/thermal/sun8i_thermal.c
+++ b/drivers/thermal/sun8i_thermal.c
@@ -66,6 +66,7 @@ struct tsensor {
 };
 
 struct ths_thermal_chip {
+	bool		has_gpadc_clk;
 	bool            has_mod_clk;
 	bool            has_bus_clk_reset;
 	bool		needs_sram;
@@ -89,6 +90,7 @@ struct ths_device {
 	struct regmap_field			*sram_regmap_field;
 	struct reset_control			*reset;
 	struct clk				*bus_clk;
+	struct clk				*gpadc_clk;
 	struct clk                              *mod_clk;
 	struct tsensor				sensor[MAX_SENSOR_NUM];
 };
@@ -417,6 +419,12 @@ static int sun8i_ths_resource_init(struct ths_device *tmdev)
 	if (ret)
 		return ret;
 
+	if (tmdev->chip->has_gpadc_clk) {
+		tmdev->gpadc_clk = devm_clk_get_enabled(&pdev->dev, "gpadc");
+		if (IS_ERR(tmdev->gpadc_clk))
+			return PTR_ERR(tmdev->gpadc_clk);
+	}
+
 	if (tmdev->chip->needs_sram) {
 		struct regmap *regmap;
 

From patchwork Sat Oct 25 04:31:26 2025
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: iuncuim <iuncuim@gmail.com>
X-Patchwork-Id: 14284215
Return-Path: 
 <linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from bombadil.infradead.org (bombadil.infradead.org
 [198.137.202.133])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by smtp.lore.kernel.org (Postfix) with ESMTPS id 54754CCD1BF
	for <linux-arm-kernel@archiver.kernel.org>;
 Sat, 25 Oct 2025 04:32:52 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=bombadil.20210309; h=Sender:List-Subscribe:List-Help
	:List-Post:List-Archive:List-Unsubscribe:List-Id:Content-Transfer-Encoding:
	MIME-Version:References:In-Reply-To:Message-ID:Date:Subject:Cc:To:From:
	Reply-To:Content-Type:Content-ID:Content-Description:Resent-Date:Resent-From:
	Resent-Sender:Resent-To:Resent-Cc:Resent-Message-ID:List-Owner;
	bh=i3+BmyLOtp7xUTMRQtQMY1TWwM+wvGI7ScMcGnjI6fc=; b=adl4Wy0zieITy3W9q50ML+HvRE
	P0pigkHpXjYUNiXBPhJuzsTeTpow6JKeUSwq2vutOquA7jpNoPZddJQRZ9BPW3YUTCr7tkRqkGDcV
	F7qQAbXmntmZFdIooq13lUd/I44ejeuJXBHyP7xvUAR6ddAOVDvV2YL7vdy1rcJiBOmoiz31C7GqK
	flf1Qga+R4dZUxAXV0TGLC1uN/nyE3XunRzMz6bQXA2dPL/LraZn2XG5bBikthn2Swkffy1TLtHch
	huMwEk88FigNbBMazaajdT3EtF2zgv3X9T8JilYMHInmBV9yFf37E0J0Fze9gSOJ4fGvyaVH6ArbN
	lafT2xBg==;
Received: from localhost ([::1] helo=bombadil.infradead.org)
	by bombadil.infradead.org with esmtp (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVxN-0000000Awce-3iVs;
	Sat, 25 Oct 2025 04:32:45 +0000
Received: from mail-lf1-x132.google.com ([2a00:1450:4864:20::132])
	by bombadil.infradead.org with esmtps (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVxL-0000000AwbH-0l4I
	for linux-arm-kernel@lists.infradead.org;
	Sat, 25 Oct 2025 04:32:44 +0000
Received: by mail-lf1-x132.google.com with SMTP id
 2adb3069b0e04-586883eb9fbso3291631e87.1
        for <linux-arm-kernel@lists.infradead.org>;
 Fri, 24 Oct 2025 21:32:42 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20230601; t=1761366761; x=1761971561;
 darn=lists.infradead.org;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:from:to:cc:subject:date
         :message-id:reply-to;
        bh=i3+BmyLOtp7xUTMRQtQMY1TWwM+wvGI7ScMcGnjI6fc=;
        b=Xt87ry3B0SPjOUIQc+32EHOEdCR6LTTXTKSE9NeNewwyIDtuCetp4WaFDY7dVw+DBB
         EyuAchXXs/PYIjsBfGTiYk5O/qU4Shyf8ROj+wS31o+PYGrONyn3ttxto3Iaurm2KVBK
         aDM3JWcMZajwZkDub+M/d5bKYjAlfxO/dP3C9mJl2lyO55vz5HsLZftokWHH0xb3TZrc
         PqWF2JgXiIkFnFtEzk7tHguxvrp8fT/Fba4mPgQBxcFW2fK0MhcRpYUXwDqVX+v23/wS
         Zfgkau49XYRtHSz5OAc7227iJiPqtQ6u9fjPtBtvWzwsWV3UkwVk25axxMRL5FSK1nQl
         HKYQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20230601; t=1761366761; x=1761971561;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:x-gm-message-state:from:to:cc
         :subject:date:message-id:reply-to;
        bh=i3+BmyLOtp7xUTMRQtQMY1TWwM+wvGI7ScMcGnjI6fc=;
        b=D12flKuluJAdTbssPLznBNIuJPLCJMj0LTpPJN+BRRGZyVCX6E0WLpSBBKS04CQ8oO
         MzQP57x4+pS4EumBA6W6yhSnyHyffNYsTH3ZMkSyUM/WyrHyPGA8/E3l3hM/Ggvbi01B
         OZxGhuC3S8tqs3b8RRN+/6vwR1Vx745O5XMESO0mQY8M/v/W1x5DYJeQW20T0ZIlDHBS
         FCtTdFeT3pL2VlzjV1gSDBLMCNbmyNYjdjmp0J+6s3Fw2yDmDaRn3o/QMIMBdZWq0owa
         vEFO9EbdfiKFP9gQ2qXb9gXuMFsSa9XpUL+yh9/SZnJBlgT4PRPpI4q9jezUb7PIxU0Q
         PjVg==
X-Forwarded-Encrypted: i=1;
 AJvYcCXxiPyhG/Vjz+RH4fnPkZh0+PNv//2it3udm1eziqP1FtUOf5ZWSQvbPy6vMoEGXKFzWIrz/5+FOLDo0Fs57ZRO@lists.infradead.org
X-Gm-Message-State: AOJu0YwePFOsC/ntzx7tWn85A9jRWHr4+3Qu0yVWHc34jjxvx9YQ82IB
	uBWqXxPMzf2LaCP5Ca981iaHO73cWWdux1lvXGHTG+xMxCWBYBeLKifN
X-Gm-Gg: ASbGncv6VAX1uDtAGvKdzp0PteeHWUZ7wQR9VUMjmkgIh2fH4BG2oD62soY1ISqKS+p
	gWKukY2wrJtANG/imkxSrfUKbS/Qvwy3Yg1A4FFRf7oFywPUaKfOCBuhoRdYF8VsDu3ng38bOr5
	50uFdbjD0DlNpIOMFRQgt0l4CHHrGexYE56yZIYOOCe5c4J5JM/j3h49+NRxRb6512rSkMf0kWt
	ZQb8SXYE6asnPyqtwA/3XPK3V3HaYQgkZ2Cl4FE2dn8PWTm4AcKWh0PA6+ok6tVSEq5XSQmJmoB
	1VxQFm652wWr31NOaT7q6BTQBKXQZ6uns2WPdPYwxrT1o8EQRvY0+8lGg5zLJIhBhpMzVqPGIn2
	r+NTsNj5pr2Txy7OY5/XQymsUjpIWyp7N8UAN/CZ6/ZMNcwA2lRPwyoa2B6ge+WEaUA==
X-Google-Smtp-Source: 
 AGHT+IEKvrwo7BMcvdLr8sMnPaVHb2OI/tgnv1kSKzQDgd1sc7jso2BrG2SWkbje2QobtPqCs0yKQw==
X-Received: by 2002:a2e:bd17:0:b0:36c:2367:b3c1 with SMTP id
 38308e7fff4ca-37797a58f18mr88205231fa.35.1761366760931;
        Fri, 24 Oct 2025 21:32:40 -0700 (PDT)
Received: from junAIR ([176.106.241.81])
        by smtp.gmail.com with ESMTPSA id
 38308e7fff4ca-378ee0ca7a0sm3409241fa.33.2025.10.24.21.32.25
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Fri, 24 Oct 2025 21:32:40 -0700 (PDT)
From: iuncuim <iuncuim@gmail.com>
To: Vasily Khoruzhick <anarsoul@gmail.com>,
	Yangtao Li <tiny.windzz@gmail.com>,
	"Rafael J. Wysocki" <rafael@kernel.org>,
	Daniel Lezcano <daniel.lezcano@linaro.org>,
	Zhang Rui <rui.zhang@intel.com>,
	Lukasz Luba <lukasz.luba@arm.com>,
	Rob Herring <robh@kernel.org>,
	Krzysztof Kozlowski <krzk+dt@kernel.org>,
	Conor Dooley <conor+dt@kernel.org>,
	Chen-Yu Tsai <wens@csie.org>,
	Jernej Skrabec <jernej.skrabec@gmail.com>,
	Samuel Holland <samuel@sholland.org>,
	Philipp Zabel <p.zabel@pengutronix.de>
Cc: Andre Przywara <andre.przywara@arm.com>,
	linux-pm@vger.kernel.org,
	devicetree@vger.kernel.org,
	linux-arm-kernel@lists.infradead.org,
	linux-sunxi@lists.linux.dev,
	linux-kernel@vger.kernel.org
Subject: [PATCH v3 3/6] thermal/drivers/sun8i: replace devm_reset_control_get
  to devm_reset_control_get_shared_deasserted
Date: Sat, 25 Oct 2025 12:31:26 +0800
Message-ID: <20251025043129.160454-4-iuncuim@gmail.com>
X-Mailer: git-send-email 2.51.0
In-Reply-To: <20251025043129.160454-1-iuncuim@gmail.com>
References: <20251025043129.160454-1-iuncuim@gmail.com>
MIME-Version: 1.0
X-CRM114-Version: 20100106-BlameMichelson ( TRE 0.8.0 (BSD) ) MR-646709E3 
X-CRM114-CacheID: sfid-20251024_213243_237750_186BC01E 
X-CRM114-Status: GOOD (  12.26  )
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.34
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: 
 <http://lists.infradead.org/mailman/options/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: 
 <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
Sender: "linux-arm-kernel" <linux-arm-kernel-bounces@lists.infradead.org>
Errors-To: 
 linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org

From: Mikhail Kalashnikov <iuncuim@gmail.com>

The A523 processor has two temperature controllers, but they share a
common reset line. Make it shared with the shared variant of
devm_reset_control_get(), and also simplify the driver by switching to
devm_reset_control_get_shared_deasserted().

Signed-off-by: Mikhail Kalashnikov <iuncuim@gmail.com>
Reviewed-by: Chen-Yu Tsai <wens@csie.org>
---
 drivers/thermal/sun8i_thermal.c | 16 +---------------
 1 file changed, 1 insertion(+), 15 deletions(-)

diff --git a/drivers/thermal/sun8i_thermal.c b/drivers/thermal/sun8i_thermal.c
index c02c398b0..aa496e1ba 100644
--- a/drivers/thermal/sun8i_thermal.c
+++ b/drivers/thermal/sun8i_thermal.c
@@ -344,11 +344,6 @@ static int sun8i_ths_calibrate(struct ths_device *tmdev)
 	return ret;
 }
 
-static void sun8i_ths_reset_control_assert(void *data)
-{
-	reset_control_assert(data);
-}
-
 static struct regmap *sun8i_ths_get_sram_regmap(struct device_node *node)
 {
 	struct platform_device *sram_pdev;
@@ -391,19 +386,10 @@ static int sun8i_ths_resource_init(struct ths_device *tmdev)
 		return PTR_ERR(tmdev->regmap);
 
 	if (tmdev->chip->has_bus_clk_reset) {
-		tmdev->reset = devm_reset_control_get(dev, NULL);
+		tmdev->reset = devm_reset_control_get_shared_deasserted(dev, NULL);
 		if (IS_ERR(tmdev->reset))
 			return PTR_ERR(tmdev->reset);
 
-		ret = reset_control_deassert(tmdev->reset);
-		if (ret)
-			return ret;
-
-		ret = devm_add_action_or_reset(dev, sun8i_ths_reset_control_assert,
-					       tmdev->reset);
-		if (ret)
-			return ret;
-
 		tmdev->bus_clk = devm_clk_get_enabled(&pdev->dev, "bus");
 		if (IS_ERR(tmdev->bus_clk))
 			return PTR_ERR(tmdev->bus_clk);

From patchwork Sat Oct 25 04:31:27 2025
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: iuncuim <iuncuim@gmail.com>
X-Patchwork-Id: 14284216
Return-Path: 
 <linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from bombadil.infradead.org (bombadil.infradead.org
 [198.137.202.133])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by smtp.lore.kernel.org (Postfix) with ESMTPS id BADE8CCF9E3
	for <linux-arm-kernel@archiver.kernel.org>;
 Sat, 25 Oct 2025 04:33:04 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=bombadil.20210309; h=Sender:List-Subscribe:List-Help
	:List-Post:List-Archive:List-Unsubscribe:List-Id:Content-Transfer-Encoding:
	MIME-Version:References:In-Reply-To:Message-ID:Date:Subject:Cc:To:From:
	Reply-To:Content-Type:Content-ID:Content-Description:Resent-Date:Resent-From:
	Resent-Sender:Resent-To:Resent-Cc:Resent-Message-ID:List-Owner;
	bh=aU3qD3/P4X3aGl+GxoTABgHAWYmZmNCs4mQEkZTLmtc=; b=ZddDN2Xj8ezkYVUD7T9GQbotGo
	vOL1xmkjb0XQaCeEMQdo7pBF+TbluRaQP8taxepE0XCq111C3jYJRWwksnONIdEi1uggJSiFa6x0J
	M4Z5oUfqm//fMtsKb7fTuTRvnSUVxmxxnG/HLjwClR5c3y/mQI/SheAbQVQDbbMLjuMNqKVmy/hR9
	PR/D6nsjQDtV8dFMTkTqY2ggvEwMC1fAgOR6p5W+2WX5buwbvEWE6+Gq7dtnrmMpq7Ok265odMWNE
	XPLCJhJO8Wo0fjPO6/PKN4axzVlBTdkbvcOYovULfeHz0rAFOk8TPQvtNdXXwxEhCYlxTSr44/zq0
	5R65bHng==;
Received: from localhost ([::1] helo=bombadil.infradead.org)
	by bombadil.infradead.org with esmtp (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVxa-0000000AwiZ-30W6;
	Sat, 25 Oct 2025 04:32:58 +0000
Received: from mail-lf1-x135.google.com ([2a00:1450:4864:20::135])
	by bombadil.infradead.org with esmtps (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVxY-0000000Awgg-3hMX
	for linux-arm-kernel@lists.infradead.org;
	Sat, 25 Oct 2025 04:32:58 +0000
Received: by mail-lf1-x135.google.com with SMTP id
 2adb3069b0e04-592f098f7adso3329162e87.0
        for <linux-arm-kernel@lists.infradead.org>;
 Fri, 24 Oct 2025 21:32:56 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20230601; t=1761366775; x=1761971575;
 darn=lists.infradead.org;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:from:to:cc:subject:date
         :message-id:reply-to;
        bh=aU3qD3/P4X3aGl+GxoTABgHAWYmZmNCs4mQEkZTLmtc=;
        b=Qp1pNLny1e2N6JjWUH6sJkl6tN6K2g39BvK9A1P0vrgTp4gs9Eo4ypkstQSGukXPBZ
         zla4sVqGy2a27SeTeSQ60QAglNdY3EF/ujSx16YsOBSzFptXxf/BSX5bGcZyhp0GNeiv
         SUw+WvqZtBSLyWW0c7ajGJeJQqDQamshMWc0e0jt88PnaopUTJkRoeRalK/UPEkW83vm
         Mmg+MQ3yG9L1fXgpAI7zfCF5NlFIH7AhGOZ9QMTZ/C43bZfP9fMnG2Iq9EGoD3BZ3Xtl
         S/VmlHs93i03HQQNK+ljNzR3wChpDVfgn/MrbibadVDC2al4m97ZW4BlXN+TMbQ0Hvd4
         qISQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20230601; t=1761366775; x=1761971575;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:x-gm-message-state:from:to:cc
         :subject:date:message-id:reply-to;
        bh=aU3qD3/P4X3aGl+GxoTABgHAWYmZmNCs4mQEkZTLmtc=;
        b=K2Bd0gPP8+A8JA334Q3qYchHB+ofcQKI7X7bC4MIQnMSvyCM/NsVitxeaqCk5pM5nP
         923DSAwEfJFxaVqBsseqjoBf9K8H6UaszQXJhPxukGVrAoWjActcHDhvnXPO7OCbJoCy
         IJu1Eb+p50y+3t1ZFLPN+5lgWSMysHHFnBR4frtNb3OLlKAca3KzjkVHVTnPl59TDxxt
         wCPGhsmK4NQki4It9z3WXSLFYRwt34wLddfo6hxU2w5PNVUhTaS1mZgAJhXB2gizqiIS
         Dvljmr9rF0bNc/XP20u7RquOz9AZy7NO4CpoMKlZnL9vgnkNZ6GZAuspI/RcEtD01mua
         a6Uw==
X-Forwarded-Encrypted: i=1;
 AJvYcCXaXpLmIOo7IfWEAf7qfYgc/OGmN1V73tHpw3WjDIizdlH7V1kramA5tIQuh0TwBjwyv7pO2AqcLEeyCcOhwgUD@lists.infradead.org
X-Gm-Message-State: AOJu0Yy0jFr/PXT6Uxtdfav0hgG0zlHd/ddzFekItm2HN3jANE3hcHuI
	LymMbAYHDplbdXO4fCMnoWV5lEsAL5Zmycdsa7dKYyNyDfahWKaihe0N
X-Gm-Gg: ASbGncvy3p8Di2x2N161qOigfdfaayc7VQ8b8w6NhdFHyRBapoGmyGE7QDcJEDfll2w
	3flKSdIoyAQBiNTKXb2v8sJQAsHymWTqVfj/skxEb2ZKtZjr0poommm6S4UFQlgLA5Dyal5yKim
	Oi8ZabvP5nGkaRcfzzwLX0pS+5V/PEs5w85JRSuktgaXrij1v2b81lSebVHo9P1YtaqafDnqUqr
	JWDm3lHOxvObj55BxOqvxG3OqaqjoQNPFVJvIieup4+cu9hvv8/iZNfihJ62+lFQIhh8E7093F3
	j5Xrp5qP7DxnxAdHUAaCqIWJbfd4W1PAJLmOqVGyGXsw3p3byMXRWCQDkamXOqQ0G5x0Wm1cQmd
	e5H5GGTW04QkPkCoixM4NOhbk4lOFiYSKcc0AF06z+Ibs5JPeo2GBNZnPDAkhsB6hhG99/h+7W3
	XW
X-Google-Smtp-Source: 
 AGHT+IH/eF1EDfMQXoqCn1NhnMmPtXp6P1wcPRtuabMRfy3u2y12SFECgEV2LqHRoicJ+PYpOzbYLg==
X-Received: by 2002:a2e:a916:0:b0:376:4430:b545 with SMTP id
 38308e7fff4ca-378e4648542mr13765001fa.49.1761366774636;
        Fri, 24 Oct 2025 21:32:54 -0700 (PDT)
Received: from junAIR ([176.106.241.81])
        by smtp.gmail.com with ESMTPSA id
 38308e7fff4ca-378ee0ca7a0sm3409241fa.33.2025.10.24.21.32.41
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Fri, 24 Oct 2025 21:32:54 -0700 (PDT)
From: iuncuim <iuncuim@gmail.com>
To: Vasily Khoruzhick <anarsoul@gmail.com>,
	Yangtao Li <tiny.windzz@gmail.com>,
	"Rafael J. Wysocki" <rafael@kernel.org>,
	Daniel Lezcano <daniel.lezcano@linaro.org>,
	Zhang Rui <rui.zhang@intel.com>,
	Lukasz Luba <lukasz.luba@arm.com>,
	Rob Herring <robh@kernel.org>,
	Krzysztof Kozlowski <krzk+dt@kernel.org>,
	Conor Dooley <conor+dt@kernel.org>,
	Chen-Yu Tsai <wens@csie.org>,
	Jernej Skrabec <jernej.skrabec@gmail.com>,
	Samuel Holland <samuel@sholland.org>,
	Philipp Zabel <p.zabel@pengutronix.de>
Cc: Andre Przywara <andre.przywara@arm.com>,
	linux-pm@vger.kernel.org,
	devicetree@vger.kernel.org,
	linux-arm-kernel@lists.infradead.org,
	linux-sunxi@lists.linux.dev,
	linux-kernel@vger.kernel.org
Subject: [PATCH v3 4/6] thermal/drivers/sun8i: get calibration data from two
 nvmem cells
Date: Sat, 25 Oct 2025 12:31:27 +0800
Message-ID: <20251025043129.160454-5-iuncuim@gmail.com>
X-Mailer: git-send-email 2.51.0
In-Reply-To: <20251025043129.160454-1-iuncuim@gmail.com>
References: <20251025043129.160454-1-iuncuim@gmail.com>
MIME-Version: 1.0
X-CRM114-Version: 20100106-BlameMichelson ( TRE 0.8.0 (BSD) ) MR-646709E3 
X-CRM114-CacheID: sfid-20251024_213256_955322_7069AB7F 
X-CRM114-Status: GOOD (  22.53  )
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.34
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: 
 <http://lists.infradead.org/mailman/options/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: 
 <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
Sender: "linux-arm-kernel" <linux-arm-kernel-bounces@lists.infradead.org>
Errors-To: 
 linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org

From: Mikhail Kalashnikov <iuncuim@gmail.com>

The A523 processor has calibration data in two nvmem cell. To be able to
add support, the ability to add data from two cells into one array must be
added.

Signed-off-by: Mikhail Kalashnikov <iuncuim@gmail.com>
---
 drivers/thermal/sun8i_thermal.c | 77 ++++++++++++++++++++++-----------
 1 file changed, 52 insertions(+), 25 deletions(-)

diff --git a/drivers/thermal/sun8i_thermal.c b/drivers/thermal/sun8i_thermal.c
index aa496e1ba..d6d8e13e5 100644
--- a/drivers/thermal/sun8i_thermal.c
+++ b/drivers/thermal/sun8i_thermal.c
@@ -303,43 +303,70 @@ static int sun50i_h6_ths_calibrate(struct ths_device *tmdev,
 
 static int sun8i_ths_calibrate(struct ths_device *tmdev)
 {
-	struct nvmem_cell *calcell;
+	struct nvmem_cell *calcell = NULL;
 	struct device *dev = tmdev->dev;
-	u16 *caldata;
-	size_t callen;
+	struct device_node *np = dev_of_node(dev);
+	struct property *prop;
+	const char *cellname;
+	u8 *caldata = NULL;
+	size_t callen = 0;
 	int ret = 0;
 
-	calcell = nvmem_cell_get(dev, "calibration");
-	if (IS_ERR(calcell)) {
-		if (PTR_ERR(calcell) == -EPROBE_DEFER)
-			return -EPROBE_DEFER;
-		/*
-		 * Even if the external calibration data stored in sid is
-		 * not accessible, the THS hardware can still work, although
-		 * the data won't be so accurate.
-		 *
-		 * The default value of calibration register is 0x800 for
-		 * every sensor, and the calibration value is usually 0x7xx
-		 * or 0x8xx, so they won't be away from the default value
-		 * for a lot.
-		 *
-		 * So here we do not return error if the calibration data is
-		 * not available, except the probe needs deferring.
-		 */
-		goto out;
+	of_property_for_each_string(np, "nvmem-cell-names", prop, cellname) {
+		size_t len;
+		u8 *caldatapart;
+
+		calcell = of_nvmem_cell_get(np, cellname);
+		if (IS_ERR(calcell)) {
+			if (PTR_ERR(calcell) == -EPROBE_DEFER)
+				return -EPROBE_DEFER;
+			/*
+			 * Even if the external calibration data stored in sid is
+			 * not accessible, the THS hardware can still work, although
+			 * the data won't be so accurate.
+			 *
+			 * The default value of calibration register is 0x800 for
+			 * every sensor, and the calibration value is usually 0x7xx
+			 * or 0x8xx, so they won't be away from the default value
+			 * for a lot.
+			 *
+			 * So here we do not return error if the calibration data is
+			 * not available, except the probe needs deferring.
+			 */
+			goto out;
+		}
+
+		caldatapart = nvmem_cell_read(calcell, &len);
+		nvmem_cell_put(calcell);
+		calcell = NULL;
+		if (IS_ERR(caldatapart)) {
+			ret = PTR_ERR(caldatapart);
+			goto out;
+		}
+
+		caldata = devm_krealloc(dev, caldata, callen + len, GFP_KERNEL);
+		if (!caldata) {
+			kfree(caldatapart);
+			ret = -ENOMEM;
+			goto out;
+		}
+
+		memcpy(caldata + callen, caldatapart, len);
+		callen += len;
+		kfree(caldatapart);
 	}
 
-	caldata = nvmem_cell_read(calcell, &callen);
 	if (IS_ERR(caldata)) {
 		ret = PTR_ERR(caldata);
 		goto out;
 	}
 
-	tmdev->chip->calibrate(tmdev, caldata, callen);
+	tmdev->chip->calibrate(tmdev, (u16 *)caldata, callen);
 
-	kfree(caldata);
+	devm_kfree(dev, caldata);
+	caldata = NULL;
 out:
-	if (!IS_ERR(calcell))
+	if (calcell && !IS_ERR(calcell))
 		nvmem_cell_put(calcell);
 	return ret;
 }

From patchwork Sat Oct 25 04:31:28 2025
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: iuncuim <iuncuim@gmail.com>
X-Patchwork-Id: 14284217
Return-Path: 
 <linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from bombadil.infradead.org (bombadil.infradead.org
 [198.137.202.133])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by smtp.lore.kernel.org (Postfix) with ESMTPS id 21AE9CCD1BF
	for <linux-arm-kernel@archiver.kernel.org>;
 Sat, 25 Oct 2025 04:33:21 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=bombadil.20210309; h=Sender:List-Subscribe:List-Help
	:List-Post:List-Archive:List-Unsubscribe:List-Id:Content-Transfer-Encoding:
	MIME-Version:References:In-Reply-To:Message-ID:Date:Subject:Cc:To:From:
	Reply-To:Content-Type:Content-ID:Content-Description:Resent-Date:Resent-From:
	Resent-Sender:Resent-To:Resent-Cc:Resent-Message-ID:List-Owner;
	bh=BpucFmLklCdqMhA9C9XyPHvfCwLqk6dMk7FGikdB0lU=; b=0D2waqgMthuss5SUvBB9PEPfz0
	B72nzQ+hB3NVSj92O4MdpMWw7nOum1qYpfQVyrNjNpfgbVFUY3lnhx4xAPn3XqVR0DJUp/Uu9cFX/
	tUZq4fvoRTzHZJn70BHa/aeLRyODo0pF8tbsW8bUEoJ6bp3yYuRc9jQSKIUbnqG1DWt/VUSz8Mraq
	G00+y+p9Nwm+YQwbvEbQaK13pQvNSD7ZhlLXKA4vFc+EVzF29CkAe2AcOGOnrKCWMtCSJiVw5cqZa
	c+Lc1rzfIf5/uzaLvaS4AUSe+ZDkeqv1cBmuV0rWo+BeCvrUew7XXgl2FulcLA4uoqS/o4M6ygvZH
	eHedY97Q==;
Received: from localhost ([::1] helo=bombadil.infradead.org)
	by bombadil.infradead.org with esmtp (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVxr-0000000AwqK-0Bh4;
	Sat, 25 Oct 2025 04:33:15 +0000
Received: from mail-lf1-x132.google.com ([2a00:1450:4864:20::132])
	by bombadil.infradead.org with esmtps (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVxn-0000000AwoZ-46ie
	for linux-arm-kernel@lists.infradead.org;
	Sat, 25 Oct 2025 04:33:13 +0000
Received: by mail-lf1-x132.google.com with SMTP id
 2adb3069b0e04-57bb7ee3142so3296299e87.0
        for <linux-arm-kernel@lists.infradead.org>;
 Fri, 24 Oct 2025 21:33:11 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20230601; t=1761366790; x=1761971590;
 darn=lists.infradead.org;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:from:to:cc:subject:date
         :message-id:reply-to;
        bh=BpucFmLklCdqMhA9C9XyPHvfCwLqk6dMk7FGikdB0lU=;
        b=GSzRr3Ow+8SDbcqqkFYmvu6qaQNoAPopVvoUIGLFg0VBzFDxpAxDAv+LRp1sRyxH8o
         etNQb4wEIhUSUzpj4Qhfcq5KL1XdHhUG+xITCIkkXV0DulamaQEvedcYmRriPUGkLZdk
         HzzJKItO5X6ZOp2RvixNWleWrvOcSsluDIbU2OXec+Y/rOlzVc39qTLfMHZbxk3Ne11O
         u5RkhqJHgYA3KV3tXTIO0WMThToONy5R7ZkPCJV2tHK1EFi3rd1q+AQpqqbPvHXhilxh
         7FtCYuXL0yrKR/YjrS0USxrKQLSS2ZjYI6YD4O9XNnEeFJRRjf/WLymt7CZVPRRZ9mDi
         J+JA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20230601; t=1761366790; x=1761971590;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:x-gm-message-state:from:to:cc
         :subject:date:message-id:reply-to;
        bh=BpucFmLklCdqMhA9C9XyPHvfCwLqk6dMk7FGikdB0lU=;
        b=Ke15yWmoN9vk0wXeUG+uDLSHKEF6zd/jTF+l5LeN6GCIsPksHx6mJy9tksO70JUBD7
         JGvlCZpWv+UZFTOmiu82Rn2Pqkv/GKb/GGHAH7cEctzERoV9JPmdQGFb35OGIZezeurv
         LdnTDEEW9Y9xBxd03V9rJC2NWoPpZUbjbCh6Xnsjjl1lxjnrc6LbjRd/V6N0electC0m
         W4Y7QwEGc5KhvH1E6BgU353rcz9GlTRRbKr/0C3hDHX8OGG7A+vkuvc9gP8ysC2xi8UQ
         70h1XpJxTlH5ONhUULFh4YQcaYEKPmx2TCm8UONzXRvDNvf+kSH1JvsM/OIEl0uwiYXj
         ALJw==
X-Forwarded-Encrypted: i=1;
 AJvYcCXGZlBXUhBmqOl9a5ufbBg2hLRSo+UxnE3hxVrnkZJyw//YPrPkdB855WsQgYmBjoStEB0xsGdfx5D3YO0nYdD/@lists.infradead.org
X-Gm-Message-State: AOJu0YwhiGuYDaXp0kWUFXrTw/5HYsOJ71hytz2tpwFsLUeVdoUVzGjP
	AFxeL1k4o2IDOA6oq+gIqbJjBANgVyX88fudWnvESPhlENNdQ4A0Pbye
X-Gm-Gg: ASbGncuKBwPIPIBWNh2SwMs0NLhL9klS3x45/zLoDGMrJG4Cv4hhOSXldMzktzD7nt1
	A8MSFkiIkz+OwujFPrDv9bV8yKPcKySrd2rOoCY00UmJrW4IFaJ1WIGi2Xgfm4vXXV0hkTohQGo
	3ZtSR6sOWicDzy5OyBviBBjV6LC0p841zOL5WmpMYKamBnx7aicXEdscrY0TzF+w9AkNcAjDyJI
	Qp59PsleuBY1OB7t8dY1h7yE6g8ZDP0O+SiYzCZzJ+3+6SaP2wCt3hXjWhcwHhMy/3IfVqfpAW0
	UALAHxz2yBAwM/xjtZdwOG09KWmJbBo+SGwn5EGatzrM+Qdsc6t+BWBEdkeM8oWTL9Slu5TQtvn
	3yEyRdssdoEiMZY5u1PjNS5EGW8+0nbpeJx1YU2I1RQj2Cq45tZmL0hUqYcaL0S0h9yNdg3jooX
	Ay
X-Google-Smtp-Source: 
 AGHT+IGI1bkGkIs+m+mcDg1QheXEpjkKTfwp/CVg0HfjhPonvg/HY7Zg24J0VSwSVIiajjlFuNgFMg==
X-Received: by 2002:a05:651c:150c:b0:378:e58f:5f10 with SMTP id
 38308e7fff4ca-378e58f613dmr12893371fa.35.1761366789881;
        Fri, 24 Oct 2025 21:33:09 -0700 (PDT)
Received: from junAIR ([176.106.241.81])
        by smtp.gmail.com with ESMTPSA id
 38308e7fff4ca-378ee0ca7a0sm3409241fa.33.2025.10.24.21.32.55
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Fri, 24 Oct 2025 21:33:09 -0700 (PDT)
From: iuncuim <iuncuim@gmail.com>
To: Vasily Khoruzhick <anarsoul@gmail.com>,
	Yangtao Li <tiny.windzz@gmail.com>,
	"Rafael J. Wysocki" <rafael@kernel.org>,
	Daniel Lezcano <daniel.lezcano@linaro.org>,
	Zhang Rui <rui.zhang@intel.com>,
	Lukasz Luba <lukasz.luba@arm.com>,
	Rob Herring <robh@kernel.org>,
	Krzysztof Kozlowski <krzk+dt@kernel.org>,
	Conor Dooley <conor+dt@kernel.org>,
	Chen-Yu Tsai <wens@csie.org>,
	Jernej Skrabec <jernej.skrabec@gmail.com>,
	Samuel Holland <samuel@sholland.org>,
	Philipp Zabel <p.zabel@pengutronix.de>
Cc: Andre Przywara <andre.przywara@arm.com>,
	linux-pm@vger.kernel.org,
	devicetree@vger.kernel.org,
	linux-arm-kernel@lists.infradead.org,
	linux-sunxi@lists.linux.dev,
	linux-kernel@vger.kernel.org
Subject: [PATCH v3 5/6] thermal/drivers/sun8i: Add support for A523 THS0/1
 controllers
Date: Sat, 25 Oct 2025 12:31:28 +0800
Message-ID: <20251025043129.160454-6-iuncuim@gmail.com>
X-Mailer: git-send-email 2.51.0
In-Reply-To: <20251025043129.160454-1-iuncuim@gmail.com>
References: <20251025043129.160454-1-iuncuim@gmail.com>
MIME-Version: 1.0
X-CRM114-Version: 20100106-BlameMichelson ( TRE 0.8.0 (BSD) ) MR-646709E3 
X-CRM114-CacheID: sfid-20251024_213312_062254_87172BD4 
X-CRM114-Status: GOOD (  26.11  )
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.34
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: 
 <http://lists.infradead.org/mailman/options/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: 
 <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
Sender: "linux-arm-kernel" <linux-arm-kernel-bounces@lists.infradead.org>
Errors-To: 
 linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org

From: Mikhail Kalashnikov <iuncuim@gmail.com>

The A523 processor has two temperature controllers, THS0 and THS1.

THS0 has only one temperature sensor, which is located in the DRAM.

THS1 does have 3 sensors:
ths1_0 - "big" cores
ths1_1 - "little" cores
ths1_2 - gpu

The datasheet mentions a fourth sensor in the NPU, but lacks any registers
for operation other than calibration registers. The vendor code reads the
value from ths1_2, but uses separate calibration data, so we get two
different values from real one.

Signed-off-by: Mikhail Kalashnikov <iuncuim@gmail.com>
---
 drivers/thermal/sun8i_thermal.c | 133 ++++++++++++++++++++++++++++++++
 1 file changed, 133 insertions(+)

diff --git a/drivers/thermal/sun8i_thermal.c b/drivers/thermal/sun8i_thermal.c
index d6d8e13e5..7d35ea3c4 100644
--- a/drivers/thermal/sun8i_thermal.c
+++ b/drivers/thermal/sun8i_thermal.c
@@ -59,6 +59,12 @@
 #define SUN50I_H6_THS_PC_TEMP_PERIOD(x)		((GENMASK(19, 0) & (x)) << 12)
 #define SUN50I_H6_THS_DATA_IRQ_STS(x)		BIT(x)
 
+#define SUN55I_A523_DELIMITER			0x7c8
+#define SUN55I_A523_OFFSET_ABOVE		2736
+#define SUN55I_A523_OFFSET_BELOW		2825
+#define SUN55I_A523_SCALE_ABOVE			74
+#define SUN55I_A523_SCALE_BELOW			65
+
 struct tsensor {
 	struct ths_device		*tmdev;
 	struct thermal_zone_device	*tzd;
@@ -116,6 +122,15 @@ static int sun50i_h5_calc_temp(struct ths_device *tmdev,
 		return -1590 * reg / 10 + 276000;
 }
 
+static int sun55i_a523_calc_temp(struct ths_device *tmdev,
+				 int id, int reg)
+{
+	if (reg >= SUN55I_A523_DELIMITER)
+		return SUN55I_A523_SCALE_ABOVE * (SUN55I_A523_OFFSET_ABOVE - reg);
+	else
+		return SUN55I_A523_SCALE_BELOW * (SUN55I_A523_OFFSET_BELOW - reg);
+}
+
 static int sun8i_ths_get_temp(struct thermal_zone_device *tz, int *temp)
 {
 	struct tsensor *s = thermal_zone_device_priv(tz);
@@ -301,6 +316,97 @@ static int sun50i_h6_ths_calibrate(struct ths_device *tmdev,
 	return 0;
 }
 
+/*
+ * The A523 nvmem calibration values. The ths1_3 is not used as it
+ * doesn't have its own sensor and doesn't have any internal switch.
+ * Instead, the value from the ths1_2 sensor is used, which gives the
+ * illusion of an independent sensor for NPU and GPU when using
+ * different calibration values.
+ *
+ * efuse layout 0x38-0x3F (caldata[0..3]):
+ *     caldata[0]      caldata[1]      caldata[2]      caldata[3]
+ * 0               16      24      32  36          48          60 64
+ * +---------------+---------------+---------------+---------------+
+ * |               |       |   temp    |  ths1_0   |  ths1_1   |   +
+ * +---------------+---------------+---------------+---------------+
+ *
+ * efuse layout 0x44-0x4B (caldata[4..7]):
+ *     caldata[4]      caldata[5]      caldata[6]      caldata[7]
+ * 0           12  16      24      32  36          48             64
+ * +---------------+---------------+---------------+---------------+
+ * |  ths1_2   |  ths1_3   |    ths0   |           |               +
+ * +---------------+---------------+---------------+---------------+
+ */
+static int sun55i_a523_ths_calibrate(struct ths_device *tmdev,
+				     u16 *caldata, int callen)
+{
+	struct device *dev = tmdev->dev;
+	int i, ft_temp;
+
+	if (!caldata[0])
+		return -EINVAL;
+
+	ft_temp = (((caldata[2] << 8) | (caldata[1] >> 8)) & FT_TEMP_MASK) * 100;
+
+	for (i = 0; i < tmdev->chip->sensor_num; i++) {
+		int sensor_reg, sensor_temp, cdata, offset;
+		/*
+		 * Chips ths0 and ths1 have common parameters for value
+		 * calibration. To separate them we can use the number of
+		 * temperature sensors on each chip.
+		 * For ths0 this value is 1.
+		 */
+		if (tmdev->chip->sensor_num == 1) {
+			sensor_reg = ((caldata[5] >> 8) | (caldata[6] << 8)) & TEMP_CALIB_MASK;
+		} else {
+			switch (i) {
+			case 0:
+				sensor_reg = (caldata[2] >> 4) & TEMP_CALIB_MASK;
+				break;
+			case 1:
+				sensor_reg = caldata[3] & TEMP_CALIB_MASK;
+				break;
+			case 2:
+				sensor_reg = caldata[4] & TEMP_CALIB_MASK;
+				break;
+			default:
+				sensor_reg = 0;
+				break;
+			}
+		}
+
+		sensor_temp = tmdev->chip->calc_temp(tmdev, i, sensor_reg);
+
+		/*
+		 * Calibration data is CALIBRATE_DEFAULT - (calculated
+		 * temperature from sensor reading at factory temperature
+		 * minus actual factory temperature) * X (scale from
+		 * temperature to register values)
+		 */
+		cdata = CALIBRATE_DEFAULT -
+			((sensor_temp - ft_temp) / SUN55I_A523_SCALE_ABOVE);
+
+		if (cdata & ~TEMP_CALIB_MASK) {
+			/*
+			 * Calibration value more than 12-bit, but calibration
+			 * register is 12-bit. In this case, ths hardware can
+			 * still work without calibration, although the data
+			 * won't be so accurate.
+			 */
+			dev_warn(dev, "sensor%d is not calibrated.\n", i);
+			continue;
+		}
+
+		offset = (i % 2) * 16;
+		regmap_update_bits(tmdev->regmap,
+				   SUN50I_H6_THS_TEMP_CALIB + (i / 2 * 4),
+				   TEMP_CALIB_MASK << offset,
+				   cdata << offset);
+	}
+
+	return 0;
+}
+
 static int sun8i_ths_calibrate(struct ths_device *tmdev)
 {
 	struct nvmem_cell *calcell = NULL;
@@ -730,6 +836,31 @@ static const struct ths_thermal_chip sun50i_h616_ths = {
 	.calc_temp = sun8i_ths_calc_temp,
 };
 
+/* The A523 has a shared reset line for both chips */
+static const struct ths_thermal_chip sun55i_a523_ths0 = {
+	.sensor_num = 1,
+	.has_bus_clk_reset = true,
+	.has_gpadc_clk = true,
+	.ft_deviation = 5000,
+	.temp_data_base = SUN50I_H6_THS_TEMP_DATA,
+	.calibrate = sun55i_a523_ths_calibrate,
+	.init = sun50i_h6_thermal_init,
+	.irq_ack = sun50i_h6_irq_ack,
+	.calc_temp = sun55i_a523_calc_temp,
+};
+
+static const struct ths_thermal_chip sun55i_a523_ths1 = {
+	.sensor_num = 3,
+	.has_bus_clk_reset = true,
+	.has_gpadc_clk = true,
+	.ft_deviation = 5000,
+	.temp_data_base = SUN50I_H6_THS_TEMP_DATA,
+	.calibrate = sun55i_a523_ths_calibrate,
+	.init = sun50i_h6_thermal_init,
+	.irq_ack = sun50i_h6_irq_ack,
+	.calc_temp = sun55i_a523_calc_temp,
+};
+
 static const struct of_device_id of_ths_match[] = {
 	{ .compatible = "allwinner,sun8i-a83t-ths", .data = &sun8i_a83t_ths },
 	{ .compatible = "allwinner,sun8i-h3-ths", .data = &sun8i_h3_ths },
@@ -740,6 +871,8 @@ static const struct of_device_id of_ths_match[] = {
 	{ .compatible = "allwinner,sun50i-h6-ths", .data = &sun50i_h6_ths },
 	{ .compatible = "allwinner,sun20i-d1-ths", .data = &sun20i_d1_ths },
 	{ .compatible = "allwinner,sun50i-h616-ths", .data = &sun50i_h616_ths },
+	{ .compatible = "allwinner,sun55i-a523-ths0", .data = &sun55i_a523_ths0 },
+	{ .compatible = "allwinner,sun55i-a523-ths1", .data = &sun55i_a523_ths1 },
 	{ /* sentinel */ },
 };
 MODULE_DEVICE_TABLE(of, of_ths_match);

From patchwork Sat Oct 25 04:31:29 2025
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: iuncuim <iuncuim@gmail.com>
X-Patchwork-Id: 14284218
Return-Path: 
 <linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from bombadil.infradead.org (bombadil.infradead.org
 [198.137.202.133])
	(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by smtp.lore.kernel.org (Postfix) with ESMTPS id 7A01BCCF9E3
	for <linux-arm-kernel@archiver.kernel.org>;
 Sat, 25 Oct 2025 04:33:35 +0000 (UTC)
DKIM-Signature: v=1; a=rsa-sha256; q=dns/txt; c=relaxed/relaxed;
	d=lists.infradead.org; s=bombadil.20210309; h=Sender:List-Subscribe:List-Help
	:List-Post:List-Archive:List-Unsubscribe:List-Id:Content-Transfer-Encoding:
	MIME-Version:References:In-Reply-To:Message-ID:Date:Subject:Cc:To:From:
	Reply-To:Content-Type:Content-ID:Content-Description:Resent-Date:Resent-From:
	Resent-Sender:Resent-To:Resent-Cc:Resent-Message-ID:List-Owner;
	bh=nGpVSnY2YGnC6StYjZi+g05NBMCnFA0794RPi3L6J+s=; b=xg9kcyKOPNHRWarGgLja6knVC2
	3zS8zItcGuJIr6ytOwIuqmKilyHMaYDV/Eo6oMUBPbENKGAukQqDL+ubmAw1cTcU68TKv5C4BFqWE
	Rq4JTQpMAQAmounJ7qKDD7lSVBcYaA80yN4Vtwi15nkx6csE+I4TOXywsQIEQK+yn8Bh5D/73DGBf
	5sJr/aRTxQuZi59TXCttlx3udwC6FUQT+SL61q/pEG47LGdHCtP0X4KUOJ0bIRwUnZNztg6JsqJy+
	9oVankVuHznbhvcxmcbLHvkikl58mdBp1L1MCiA4k87/UwR8cOVygSxoPmdLFZNFdV1xlN+RqSALr
	HwUU6WvQ==;
Received: from localhost ([::1] helo=bombadil.infradead.org)
	by bombadil.infradead.org with esmtp (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVy5-0000000Awxm-1vMy;
	Sat, 25 Oct 2025 04:33:29 +0000
Received: from mail-lj1-x233.google.com ([2a00:1450:4864:20::233])
	by bombadil.infradead.org with esmtps (Exim 4.98.2 #2 (Red Hat Linux))
	id 1vCVy2-0000000Aww5-1XdE
	for linux-arm-kernel@lists.infradead.org;
	Sat, 25 Oct 2025 04:33:27 +0000
Received: by mail-lj1-x233.google.com with SMTP id
 38308e7fff4ca-378d54f657fso33856981fa.2
        for <linux-arm-kernel@lists.infradead.org>;
 Fri, 24 Oct 2025 21:33:25 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20230601; t=1761366804; x=1761971604;
 darn=lists.infradead.org;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:from:to:cc:subject:date
         :message-id:reply-to;
        bh=nGpVSnY2YGnC6StYjZi+g05NBMCnFA0794RPi3L6J+s=;
        b=BAlM6Thd74x4FhMCaiE3NPzCDvv6cyLe+jmiqHakrwW+XbrVWecPAuhimiGLsgPTQ+
         t2fSUW4nO5I3dilAHQtzcSeQi/5hQVkBYv+rWUyegfxWXRYrYTqj2xl5xhi8PnjwEweH
         5DETXK85gvFtSd3PLiuakabGBmMQhhFUs0tALFSHaP/9hl+IhmfRiXcjZmiytj7PIst8
         x3QTvFmBJVaCRsniB7l0tySkV2VxoAA8cICb5gGB9v+e9YYr97w/t1ZNFqUpwoBZy9Z/
         EolWyA0qAG3jVuEVL0MjD1sooUOoVssfKLkArl06PtrwAY5YG8LcdJisH1dSOxz1t6wV
         u1Hg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20230601; t=1761366804; x=1761971604;
        h=content-transfer-encoding:mime-version:references:in-reply-to
         :message-id:date:subject:cc:to:from:x-gm-message-state:from:to:cc
         :subject:date:message-id:reply-to;
        bh=nGpVSnY2YGnC6StYjZi+g05NBMCnFA0794RPi3L6J+s=;
        b=Njr0sF3Th1mB0Jv5dwIILlqfQ8rsH+sWPzen/OBkShM18n0ChuyFL85w1nGJmRhi3K
         vBnzevGSuTIWGA4S6Qz6SDNgu1Z1bl+6qpPNZJMtsqPHwl0PR+3h8TFZdKxj4ob688nG
         Nx71dd/W/RTkhEuqgyuSL1o0a5GVkyWk+OLj6em179xE2B0NzQIwhPa2bW/plZduwNfo
         bxXZgg7lbVpMx5/NTq9Z1tweNNNsdj+wx3d3aSRvWUfkafxTwthaYSciIjUa+tBZCrss
         wDOtihHBbTIwSXd7TrWyVd73HJVcBGW3Zt5yX/APCiRCULR4xP6lKCbYs6qnmFRV9+Xl
         71bQ==
X-Forwarded-Encrypted: i=1;
 AJvYcCWkSuPNvuRwZvz7ZASZ76093WHDtMLNm/rR0IIjwfFkx2k30kipFhwqwSX7/VN1b0+xQf0bWbn1CJZ/b1DE+tui@lists.infradead.org
X-Gm-Message-State: AOJu0YxFjZxQWRBWKS5twfBKnlmpcIkCyNwf4+SktOQxH8D3OOMJEIQE
	RRadW25l5HxAN24bByDsOdhqYfzLn/Qc6sRK9Dbc7Hvq1jG69nS/9Nuv
X-Gm-Gg: ASbGnctnATIaEjZGY1zmw6vtsu2+vSMgFIM1derrvox6V6ZaMjR36+20WJ4kpDl/JbN
	Hh5vQLVr7GKoKFwSThu0by03t7Amb6aDyv8zyiA8085O9Gn+OqoBSVTDiMHG6BMUXs8AiNT+TMX
	Lg3n9JXg9mMnOqBe5pbMKokAfV4NGREZJ48EsTUxcCUJ70cKhIIe8V3tvVycpc80URNO6FMP6l4
	BkGPCuMohjDS4ZpP6BJQ1gMJeBkuk/w5LMsUPL+7Y4Aa9u3kRIJluKOPUyAqIcLUgnblfrnpAUn
	XC6Mm3nOMVLxLwBKYdHy67bJVhphGms39TGsJCljEWp+n75l6yaRawKszs1+kaS4Ih9bbeEVq6K
	GtiNylXzuAJAqWKkqQA+O1cyck+Vf2JdCNB936URo+KQ2nB27Yq5gPxBFB1m77OV8w+DqbK+vml
	Hy
X-Google-Smtp-Source: 
 AGHT+IGj8aCrZj29Rg9HWwRsU0X1CDkriVFw8huU9nZ548l08naKsaur1apTfOMiV+X0MeyojMrrVA==
X-Received: by 2002:a2e:b8cc:0:b0:36a:925e:cf3c with SMTP id
 38308e7fff4ca-37797a39146mr98679831fa.31.1761366803946;
        Fri, 24 Oct 2025 21:33:23 -0700 (PDT)
Received: from junAIR ([176.106.241.81])
        by smtp.gmail.com with ESMTPSA id
 38308e7fff4ca-378ee0ca7a0sm3409241fa.33.2025.10.24.21.33.11
        (version=TLS1_3 cipher=TLS_AES_256_GCM_SHA384 bits=256/256);
        Fri, 24 Oct 2025 21:33:23 -0700 (PDT)
From: iuncuim <iuncuim@gmail.com>
To: Vasily Khoruzhick <anarsoul@gmail.com>,
	Yangtao Li <tiny.windzz@gmail.com>,
	"Rafael J. Wysocki" <rafael@kernel.org>,
	Daniel Lezcano <daniel.lezcano@linaro.org>,
	Zhang Rui <rui.zhang@intel.com>,
	Lukasz Luba <lukasz.luba@arm.com>,
	Rob Herring <robh@kernel.org>,
	Krzysztof Kozlowski <krzk+dt@kernel.org>,
	Conor Dooley <conor+dt@kernel.org>,
	Chen-Yu Tsai <wens@csie.org>,
	Jernej Skrabec <jernej.skrabec@gmail.com>,
	Samuel Holland <samuel@sholland.org>,
	Philipp Zabel <p.zabel@pengutronix.de>
Cc: Andre Przywara <andre.przywara@arm.com>,
	linux-pm@vger.kernel.org,
	devicetree@vger.kernel.org,
	linux-arm-kernel@lists.infradead.org,
	linux-sunxi@lists.linux.dev,
	linux-kernel@vger.kernel.org
Subject: [PATCH v3 6/6] Allwinner: A523: add support for A523 THS0/1
 controllers
Date: Sat, 25 Oct 2025 12:31:29 +0800
Message-ID: <20251025043129.160454-7-iuncuim@gmail.com>
X-Mailer: git-send-email 2.51.0
In-Reply-To: <20251025043129.160454-1-iuncuim@gmail.com>
References: <20251025043129.160454-1-iuncuim@gmail.com>
MIME-Version: 1.0
X-CRM114-Version: 20100106-BlameMichelson ( TRE 0.8.0 (BSD) ) MR-646709E3 
X-CRM114-CacheID: sfid-20251024_213326_449478_EFD1EBFE 
X-CRM114-Status: GOOD (  13.18  )
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.34
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: 
 <http://lists.infradead.org/mailman/options/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: 
 <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>,
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
Sender: "linux-arm-kernel" <linux-arm-kernel-bounces@lists.infradead.org>
Errors-To: 
 linux-arm-kernel-bounces+linux-arm-kernel=archiver.kernel.org@lists.infradead.org

From: Mikhail Kalashnikov <iuncuim@gmail.com>

The A523 processor has two temperature controllers, THS0 and THS1.
THS0 has only one temperature sensor, which is located in the DRAM.

THS1 does have 3 sensors:
ths1_0 - "big" cores
ths1_1 - "little" cores
ths1_2 - gpu

Add the thermal sensor configuration and the thermal zones.
Trips temperature, polling-delay and sustainable-power parameters are
derived from the manufacturer's BSP.

Signed-off-by: Mikhail Kalashnikov <iuncuim@gmail.com>
---
 .../arm64/boot/dts/allwinner/sun55i-a523.dtsi | 154 ++++++++++++++++++
 1 file changed, 154 insertions(+)

diff --git a/arch/arm64/boot/dts/allwinner/sun55i-a523.dtsi b/arch/arm64/boot/dts/allwinner/sun55i-a523.dtsi
index 7b36c47a3..0cbe73601 100644
--- a/arch/arm64/boot/dts/allwinner/sun55i-a523.dtsi
+++ b/arch/arm64/boot/dts/allwinner/sun55i-a523.dtsi
@@ -11,6 +11,7 @@
 #include <dt-bindings/reset/sun55i-a523-r-ccu.h>
 #include <dt-bindings/power/allwinner,sun55i-a523-ppu.h>
 #include <dt-bindings/power/allwinner,sun55i-a523-pck-600.h>
+#include <dt-bindings/thermal/thermal.h>
 
 / {
 	interrupt-parent = <&gic>;
@@ -26,6 +27,7 @@ cpu0: cpu@0 {
 			device_type = "cpu";
 			reg = <0x000>;
 			enable-method = "psci";
+			#cooling-cells = <2>;
 		};
 
 		cpu1: cpu@100 {
@@ -33,6 +35,7 @@ cpu1: cpu@100 {
 			device_type = "cpu";
 			reg = <0x100>;
 			enable-method = "psci";
+			#cooling-cells = <2>;
 		};
 
 		cpu2: cpu@200 {
@@ -40,6 +43,7 @@ cpu2: cpu@200 {
 			device_type = "cpu";
 			reg = <0x200>;
 			enable-method = "psci";
+			#cooling-cells = <2>;
 		};
 
 		cpu3: cpu@300 {
@@ -47,6 +51,7 @@ cpu3: cpu@300 {
 			device_type = "cpu";
 			reg = <0x300>;
 			enable-method = "psci";
+			#cooling-cells = <2>;
 		};
 
 		cpu4: cpu@400 {
@@ -54,6 +59,7 @@ cpu4: cpu@400 {
 			device_type = "cpu";
 			reg = <0x400>;
 			enable-method = "psci";
+			#cooling-cells = <2>;
 		};
 
 		cpu5: cpu@500 {
@@ -61,6 +67,7 @@ cpu5: cpu@500 {
 			device_type = "cpu";
 			reg = <0x500>;
 			enable-method = "psci";
+			#cooling-cells = <2>;
 		};
 
 		cpu6: cpu@600 {
@@ -68,6 +75,7 @@ cpu6: cpu@600 {
 			device_type = "cpu";
 			reg = <0x600>;
 			enable-method = "psci";
+			#cooling-cells = <2>;
 		};
 
 		cpu7: cpu@700 {
@@ -75,6 +83,7 @@ cpu7: cpu@700 {
 			device_type = "cpu";
 			reg = <0x700>;
 			enable-method = "psci";
+			#cooling-cells = <2>;
 		};
 	};
 
@@ -398,12 +407,46 @@ syscon: syscon@3000000 {
 			ranges;
 		};
 
+		ths1: thermal-sensor@2009400 {
+			compatible = "allwinner,sun55i-a523-ths1";
+			reg = <0x02009400 0x400>;
+			interrupts = <GIC_SPI 62 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&ccu CLK_BUS_THS>, <&ccu CLK_GPADC1>;
+			clock-names = "bus", "gpadc";
+			resets = <&ccu RST_BUS_THS>;
+			nvmem-cells = <&ths_calibration0>, <&ths_calibration1>;
+			nvmem-cell-names = "calibration",
+				     "calibration-second-part";
+			#thermal-sensor-cells = <1>;
+		};
+
+		ths0: thermal-sensor@200a000 {
+			compatible = "allwinner,sun55i-a523-ths0";
+			reg = <0x0200a000 0x400>;
+			interrupts = <GIC_SPI 39 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&ccu CLK_BUS_THS>, <&ccu CLK_GPADC0>;
+			clock-names = "bus", "gpadc";
+			resets = <&ccu RST_BUS_THS>;
+			nvmem-cells = <&ths_calibration0>, <&ths_calibration1>;
+			nvmem-cell-names = "calibration",
+				     "calibration-second-part";
+			#thermal-sensor-cells = <0>;
+		};
+
 		sid: efuse@3006000 {
 			compatible = "allwinner,sun55i-a523-sid",
 				     "allwinner,sun50i-a64-sid";
 			reg = <0x03006000 0x1000>;
 			#address-cells = <1>;
 			#size-cells = <1>;
+
+			ths_calibration0: ths-calibration0@38 {
+				reg = <0x38 0x8>;
+			};
+
+			ths_calibration1: ths-calibration1@44 {
+				reg = <0x44 0x8>;
+			};
 		};
 
 		gic: interrupt-controller@3400000 {
@@ -732,4 +775,115 @@ npu: npu@7122000 {
 			power-domains = <&ppu PD_NPU>;
 		};
 	};
+
+	thermal-zones {
+		cpu0_thermal: cpu0-thermal {
+			polling-delay-passive = <100>;
+			polling-delay = <1000>;
+			thermal-sensors = <&ths1 1>;
+			sustainable-power = <1200>;
+
+			trips {
+				cpu0_threshold: cpu-trip-0 {
+					temperature = <70000>;
+					type = "passive";
+					hysteresis = <0>;
+				};
+				cpu0_target: cpu-trip-1 {
+					temperature = <90000>;
+					type = "passive";
+					hysteresis = <0>;
+				};
+				cpu0_critical: cpu-trip-2 {
+					temperature = <110000>;
+					type = "critical";
+					hysteresis = <0>;
+				};
+			};
+
+			cooling-maps {
+				map0 {
+					trip = <&cpu0_target>;
+					cooling-device = <&cpu0 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+							 <&cpu1 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+							 <&cpu2 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+							 <&cpu3 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+				};
+			};
+		};
+
+		cpu4_thermal: cpu4-thermal {
+			polling-delay-passive = <100>;
+			polling-delay = <1000>;
+			thermal-sensors = <&ths1 0>;
+			sustainable-power = <1600>;
+
+			trips {
+				cpu4_threshold: cpu-trip-0 {
+					temperature = <70000>;
+					type = "passive";
+					hysteresis = <0>;
+				};
+				cpu4_target: cpu-trip-1 {
+					temperature = <90000>;
+					type = "passive";
+					hysteresis = <0>;
+				};
+				cpu4_critical: cpu-trip-2 {
+					temperature = <110000>;
+					type = "critical";
+					hysteresis = <0>;
+				};
+			};
+
+			cooling-maps {
+				map0 {
+					trip = <&cpu4_target>;
+					cooling-device = <&cpu4 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+							 <&cpu5 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+							 <&cpu6 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+							 <&cpu7 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+				};
+			};
+		};
+
+		gpu-thermal {
+			polling-delay-passive = <100>;
+			polling-delay = <1000>;
+			thermal-sensors = <&ths1 2>;
+			sustainable-power = <2400>;
+
+			gpu-trips {
+				gpu_temp_threshold: gpu-trip-0 {
+					temperature = <60000>;
+					type = "passive";
+					hysteresis = <0>;
+				};
+				gpu_temp_target: gpu-trip-1 {
+					temperature = <90000>;
+					type = "passive";
+					hysteresis = <0>;
+				};
+				gpu_temp_critical: gpu-trip-2 {
+					temperature = <110000>;
+					type = "critical";
+					hysteresis = <0>;
+				};
+			};
+		};
+
+		ddr-thermal {
+			polling-delay-passive = <0>;
+			polling-delay = <0>;
+			thermal-sensors = <&ths0>;
+
+			trips {
+				ddr_temp_critical: ddr-trip-0 {
+					temperature = <110000>;
+					type = "critical";
+					hysteresis = <0>;
+				};
+			};
+		};
+	};
 };
